lager/lenses/tuple.hpp

```diff
@@ -0,0 +1,194 @@
+#pragma once
+
+#include <tuple>
+#include <type_traits>
+
+#include <lager/lenses.hpp>
+#include <lager/lenses/attr.hpp>
+
+namespace lager {
+namespace lenses {
+namespace detail {
+
+template <typename T>
+struct templated_type_contructor;
+
+template <template <typename...> typename T, typename... Params>
+struct templated_type_contructor<T<Params...>>
+{
+    template <typename... Ts>
+    static auto construct(Ts&&... ps)
+    {
+        return T<std::remove_reference_t<Ts>...>{std::forward<Ts>(ps)...};
+    }
+};
+
+template <size_t I, typename... Tuples>
+decltype(auto) multiget(Tuples&&... tuples)
+{
+    return std::forward_as_tuple(std::get<I>(std::forward<Tuples>(tuples))...);
+};
+
+template <typename ResType, typename F, typename... Tuples, size_t... Seq>
+decltype(auto) apply_zip_impl(std::index_sequence<Seq...>, F&& f, Tuples&&... t)
+{
+    return templated_type_contructor<ResType>::construct(std::apply(
+        std::forward<F>(f), multiget<Seq>(std::forward<Tuples>(t)...))...);
+}
+
+template <typename ResType, typename F, typename... Tuple>
+decltype(auto) apply_zip(F&& f, Tuple&&... t)
+{
+    constexpr size_t len =
+        std::min({std::tuple_size_v<std::remove_reference_t<Tuple>>...});
+    return apply_zip_impl<ResType>(std::make_index_sequence<len>{},
+                                   std::forward<F>(f),
+                                   std::forward<Tuple>(t)...);
+}
+
+namespace fold_op {
+template <typename Init, typename Elem>
+decltype(auto) operator+(Init&& init, Elem&& elem)
+{
+    auto&& [lens, part] = LAGER_FWD(elem);
+    return ::lager::set(LAGER_FWD(lens), LAGER_FWD(init), LAGER_FWD(part));
+}
+} // namespace fold_op
+
+template <typename Whole, typename LensTuple, typename PartTuple, size_t... Seq>
+decltype(auto) set_fold_impl(Whole&& whole,
+                             LensTuple&& lenses,
+                             PartTuple&& parts,
+                             std::index_sequence<Seq...>)
+{
+    using fold_op::operator+;
+    return (LAGER_FWD(whole) + ... +
+            multiget<Seq>(LAGER_FWD(lenses), LAGER_FWD(parts)));
+}
+
+template <typename Whole, typename LensTuple, typename PartTuple>
+decltype(auto) set_fold(Whole&& whole, LensTuple&& lenses, PartTuple&& parts)
+{
+    return set_fold_impl(
+        LAGER_FWD(whole),
+        LAGER_FWD(lenses),
+        LAGER_FWD(parts),
+        std::make_index_sequence<
+            std::tuple_size_v<std::remove_reference_t<LensTuple>>>());
+}
+
+// note: has to be const & to avoid potential moved from UB
+template <class T, size_t... Seq>
+auto dup_impl(T const& x, std::index_sequence<Seq...>)
+{
+    return std::tuple<decltype(static_cast<void>(Seq), x)...>(
```

Comment:
Nice use of `static_cast<void>` for safety. Could also be `(void)x`?

Reply:
yeah but a lot of tools (that includes some code reviewers) will give warnings for C style casts, even for `(void)x`.

Reply:
Damn tools... :p